sklearn/cluster/_kmeans.py

```diff
@@ -1057,7 +1057,9 @@ def fit(self, X, y=None, sample_weight=None):
                 x_squared_norms=x_squared_norms, n_threads=self._n_threads)
 
             # determine if these results are the best so far
-            if best_inertia is None or inertia < best_inertia:
+            # allow small tolerance on the inertia to accommodate for
+            # non-deterministic rounding errors due to parallel computation
+            if best_inertia is None or inertia < best_inertia * (1 - 1e-6):
```

Comment:
Precising that this is just above the float32 eps [as you mentioned](https://github.com/scikit-learn/scikit-learn/pull/20200#issuecomment-856791625) might be informative here.